rust/src/detect/transform_base64.rs

```diff
@@ -0,0 +1,449 @@
+     * the results would be as follows. See the unittest B64TestVectorsRFC2045 in
+     * src/util-base64.c
+     *
+     * BASE64("") = ""
+     * BASE64("f") = "Zg=="
+     * BASE64("fo") = "Zm8="
+     * BASE64("foo") = "Zm9v"
+     * BASE64("foob") = "Zm9vYg=="
+     * BASE64("fooba") = "Zm9vYmE="
+     * BASE64("foobar") = "Zm9vYmFy"
+     * BASE64("foobar") = "Zm 9v Ym Fy"   <-- Notice how the spaces are ignored
+     * BASE64("foobar") = "Zm$9vYm.Fy"    # According to RFC 2045, All line breaks or *other
+     * characters* not found in base64 alphabet must be ignored by decoding software
+     * */
+    ModeRFC2045, /* SPs are allowed during transfer but must be skipped by Decoder */
+    ModeStrict,
+    /* If the following strings were to be passed to the decoder with RFC4648 mode,
+     * the results would be as follows. See the unittest B64TestVectorsRFC4648 in
+     * src/util-base64.c
+     *
+     * BASE64("") = ""
+     * BASE64("f") = "Zg=="
+     * BASE64("fo") = "Zm8="
+     * BASE64("foo") = "Zm9v"
+     * BASE64("foob") = "Zm9vYg=="
+     * BASE64("fooba") = "Zm9vYmE="
+     * BASE64("foobar") = "Zm9vYmFy"
+     * BASE64("f") = "Zm 9v Ym Fy"   <-- Notice how the processing stops once space is encountered
+     * BASE64("f") = "Zm$9vYm.Fy"    <-- Notice how the processing stops once an invalid char is
+     * encountered
+     * */
+    ModeRFC4648, /* reject the encoded data if it contains characters outside the base alphabet */
+}
+
+pub const TRANSFORM_FROM_BASE64_MODE_DEFAULT: DetectBase64Mode = DetectBase64Mode::ModeRFC4648;
+
+const DETECT_TRANSFORM_BASE64_MAX_PARAM_COUNT: usize = 3;
+pub const DETECT_TRANSFORM_BASE64_FLAG_MODE: u8 = 0x01;
+pub const DETECT_TRANSFORM_BASE64_FLAG_NBYTES: u8 = 0x02;
+pub const DETECT_TRANSFORM_BASE64_FLAG_OFFSET: u8 = 0x04;
+pub const DETECT_TRANSFORM_BASE64_FLAG_OFFSET_VAR: u8 = 0x08;
+pub const DETECT_TRANSFORM_BASE64_FLAG_NBYTES_VAR: u8 = 0x10;
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct SCDetectTransformFromBase64Data {
+    flags: u8,
+    nbytes: u32,
+    nbytes_str: *const c_char,
+    offset: u32,
+    offset_str: *const c_char,
+    mode: DetectBase64Mode,
+}
+
+impl Drop for SCDetectTransformFromBase64Data {
+    fn drop(&mut self) {
+        unsafe {
+            if !self.offset_str.is_null() {
+                let _ = CString::from_raw(self.offset_str as *mut c_char);
+            }
+            if !self.nbytes_str.is_null() {
+                let _ = CString::from_raw(self.nbytes_str as *mut c_char);
+            }
+        }
+    }
+}
+impl Default for SCDetectTransformFromBase64Data {
+    fn default() -> Self {
+        SCDetectTransformFromBase64Data {
+            flags: 0,
+            nbytes: 0,
+            nbytes_str: std::ptr::null_mut(),
+            offset: 0,
+            offset_str: std::ptr::null_mut(),
+            mode: TRANSFORM_FROM_BASE64_MODE_DEFAULT,
+        }
+    }
+}
+
+impl SCDetectTransformFromBase64Data {
+    pub fn new() -> Self {
```

Comment:
Why do we need `new` ? Can we not just use default ?

Reply:
Idiomatically you have both..  `new` will show up in the docs under the struct, not under the traits section making it more visible.  If new exists with no arguments, default typically calls it.  Default then allows for nice stuff elsewhere, like a default instance in serde for instance. So it is common to have both, but not required.

Reply:
I'm going to leave this as is.